lib/compat/wordpress-6.9/class-gutenberg-rest-attachments-controller.php

```diff
@@ -0,0 +1,506 @@
+<?php
+/**
+ * REST API: Gutenberg_REST_Attachments_Controller class
+ *
+ * @package gutenberg
+ * @since 6.9.0
+ * @see WP_REST_Attachments_Controller
+ */
+
+class Gutenberg_REST_Attachments_Controller extends WP_REST_Attachments_Controller {
+	/**
+	 * Registers the routes for attachments.
+	 *
+	 * @see register_rest_route()
+	 */
+	public function register_routes(): void {
+		parent::register_routes();
+	}
+	/**
+	 * Applies edits to a media item and creates a new attachment record.
+	 *
+	 * @since 5.5.0
+	 * @since 6.9.0 Adds flips capability and editable fields for the newly-created attachment post.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, WP_Error object on failure.
+	 */
+	public function edit_media_item( $request ) {
+		require_once ABSPATH . 'wp-admin/includes/image.php';
+
+		$attachment_id = $request['id'];
+
+		// This also confirms the attachment is an image.
+		$image_file = wp_get_original_image_path( $attachment_id );
+		$image_meta = wp_get_attachment_metadata( $attachment_id );
+
+		if (
+			! $image_meta ||
+			! $image_file ||
+			! wp_image_file_matches_image_meta( $request['src'], $image_meta, $attachment_id )
+		) {
+			return new WP_Error(
+				'rest_unknown_attachment',
+				__( 'Unable to get meta information for file.', 'gutenberg' ),
+				array( 'status' => 404 )
+			);
+		}
+
+		$supported_types = array( 'image/jpeg', 'image/png', 'image/gif', 'image/webp', 'image/avif', 'image/heic' );
+		$mime_type       = get_post_mime_type( $attachment_id );
+		if ( ! in_array( $mime_type, $supported_types, true ) ) {
+			return new WP_Error(
+				'rest_cannot_edit_file_type',
+				__( 'This type of file cannot be edited.', 'gutenberg' ),
+				array( 'status' => 400 )
+			);
+		}
+
+		// The `modifiers` param takes precedence over the older format.
+		if ( isset( $request['modifiers'] ) ) {
+			$modifiers = $request['modifiers'];
+		} else {
+			$modifiers = array();
+
+			if ( ! empty( $request['flip'] ) ) {
```

Comment:
I noticed that further down we're expecting both the horizontal and vertical keys to be available within `flip`.
Do we need to check that `horizontal` and `vertical` are set within it? I.e. something like the following rather than a `! empty` check?:

```suggestion
			if ( isset( $request['flip']['horizontal'], $request['flip']['vertical'] ) ) {
```

Reply:
Good idea. I'm not sure we need to check for both. I think it can be either or technically. 

In that case I probably need to then tighten up the flip operation below

> $result = $image_editor->flip( 0 !== (int) $args['flip']['vertical'], 0 !== (int) $args['flip']['horizontal'] );

Reply:
> Good idea. I'm not sure we need to check for both. I think it can be either or technically.

Ah, true! Yeah, whichever way works 👍

Reply:
Looks like the Image Classes don't require both, e.g., https://github.com/WordPress/wordpress-develop/blob/6.8.2/src/wp-includes/class-wp-image-editor-imagick.php#L761


I'll update this change and add a couple more tests.

Thanks!